File: backend/TirthSeva.API/Controllers/PaymentsController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using System.Security.Claims;
using TirthSeva.API.Models;
using TirthSeva.API.Services;

namespace TirthSeva.API.Controllers
{
    [Authorize]
    [Route("api/[controller]")]
    [ApiController]
    public class PaymentsController : ControllerBase
    {
        private readonly PaymentService _paymentService;

        public PaymentsController(PaymentService paymentService)
        {
            _paymentService = paymentService;
        }

        [HttpPost("process")]
        public async Task<ActionResult<Payment>> ProcessPayment([FromBody] ProcessPaymentRequest request)
        {
            var userIdClaim = User.FindFirst(ClaimTypes.NameIdentifier);
            if (userIdClaim == null)
            {
                return Unauthorized();
            }

            var userId = int.Parse(userIdClaim.Value);
            var payment = await _paymentService.ProcessPaymentAsync(request.BookingId, request.PaymentMethod, userId);
            
            if (payment == null)
            {
                return BadRequest(new { message = "Payment failed. Booking not found or already paid." });
            }

            return Ok(payment);
        }

        [HttpPost("verify")]
        public async Task<ActionResult> VerifyPayment([FromBody] VerifyPaymentRequest request)
        {
            var result = await _paymentService.VerifyPaymentAsync(request.TransactionId);
            
            if (!result)
            {
                return NotFound(new { message = "Payment not found or failed" });
            }

            return Ok(new { message = "Payment verified successfully", status = "Success" });
        }

        [HttpGet("history")]
        public async Task<ActionResult<List<Payment>>> GetPaymentHistory()
        {
            var userIdClaim = User.FindFirst(ClaimTypes.NameIdentifier);
            if (userIdClaim == null)
            {
                return Unauthorized();
            }

            var userId = int.Parse(userIdClaim.Value);
            var payments = await _paymentService.GetPaymentHistoryAsync(userId);
            return Ok(payments);
        }

        [HttpGet("booking/{bookingId}")]
        public async Task<ActionResult<Payment>> GetPaymentByBooking(int bookingId)
        {
            var payment = await _paymentService.GetPaymentByBookingIdAsync(bookingId);
            
            if (payment == null)
            {
                return NotFound();
            }

            return Ok(payment);
        }
    }

    public class ProcessPaymentRequest
    {
        public int BookingId { get; set; }
        public string PaymentMethod { get; set; } = "Mock";
    }

    public class VerifyPaymentRequest
    {
        public string TransactionId { get; set; } = string.Empty;
    }
}

File: backend-spring/src/main/java/com/tirthseva/api/controller/PaymentsController.java
package com.tirthseva.api.controller;

import com.tirthseva.api.dto.payment.ProcessPaymentRequest;
import com.tirthseva.api.dto.payment.VerifyPaymentRequest;
import com.tirthseva.api.entity.Payment;
import com.tirthseva.api.entity.User;
import com.tirthseva.api.service.PaymentService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.annotation.AuthenticationPrincipal;
import org.springframework.web.bind.annotation.*;

import java.util.List;
import java.util.Map;

@RestController
@RequestMapping("/api/payments")
@RequiredArgsConstructor
public class PaymentsController {

    private final PaymentService paymentService;

    @PostMapping("/process")
    public ResponseEntity<?> processPayment(@RequestBody ProcessPaymentRequest request,
            @AuthenticationPrincipal User user) {
        if (user == null) {
            return ResponseEntity.status(401).build();
        }

        Payment payment = paymentService.processPayment(request.getBookingId(), request.getPaymentMethod(),
                user.getId());

        if (payment == null) {
            return ResponseEntity.badRequest()
                    .body(Map.of("message", "Payment failed. Booking not found or already paid."));
        }

        return ResponseEntity.ok(payment);
    }

    @PostMapping("/verify")
    public ResponseEntity<?> verifyPayment(@RequestBody VerifyPaymentRequest request) {
        boolean result = paymentService.verifyPayment(request.getTransactionId());

        if (!result) {
            return ResponseEntity.status(404)
                    .body(Map.of("message", "Payment not found or failed"));
        }

        return ResponseEntity.ok(Map.of("message", "Payment verified successfully", "status", "Success"));
    }

    @GetMapping("/history")
    public ResponseEntity<List<Payment>> getPaymentHistory(@AuthenticationPrincipal User user) {
        if (user == null) {
            return ResponseEntity.status(401).build();
        }

        List<Payment> payments = paymentService.getPaymentHistory(user.getId());
        return ResponseEntity.ok(payments);
    }

    @GetMapping("/booking/{bookingId}")
    public ResponseEntity<Payment> getPaymentByBooking(@PathVariable Integer bookingId) {
        Payment payment = paymentService.getPaymentByBookingId(bookingId);

        if (payment == null) {
            return ResponseEntity.notFound().build();
        }

        return ResponseEntity.ok(payment);
    }
}
